File: data/granger_ts.jl
@time using LinearAlgebra
@time using Statistics
@time using DataFrames
@time using CSV
@time using MarSwitching
@time using GLM
@time using BenchmarkTools

replace(x, to) = ismissing(x) ? to : x

data = CSV.read("data/df_rets_granger.csv", DataFrame)[:, 2:end]
data = Matrix(replace.(data, Inf))


function na_share(df::Matrix{Float64})
    return sum(isinf.(df), dims=1) ./ size(df, 1)
end

function remove_infs(df::Matrix{Float64})
    
    infs = []
    for i in 1:size(df, 2)
        infs = [infs; findall(isinf.(df[:,i]))]
    end

    return df[(1:end) .∉ (infs,),:]
end

function std_err(X::Matrix, y::Vector, β::Vector)
    mse = mean((y .- X * β).^2)
    σ² = mse * (X'X)^(-1)
    return sqrt.(diag(σ²))
end    

function preproc_df(df::Matrix{Float64})
    pair = [df[2:end, 1] add_lags(df[:, 1], 1)[:,2] add_lags(df[:, 2], 1)[:,2]]
    pair = remove_infs(pair)
    pair = [pair ones(size(pair, 1))]
    return pair
end 


function granger_cause(df::Matrix{Float64})
    pair = preproc_df(df)

    X = pair[:, 2:end]
    y = pair[:, 1]
    
    β = (X'X) \ X'y
    Σ = std_err(X, y, β)
    
    return β[2], abs(β[2] / Σ[2]) > 1.96
end


function granger_conect(df::Matrix{Float64})
    granger_mat = zeros(size(df, 2), size(df, 2))

    for i in 1:size(df, 2)
        for j in 1:size(df, 2)
            if i == j
                granger_mat[i, j] = 0
                continue
            end

            if any(na_share(df[:, [i, j]]) .>= 0.9)
                granger_mat[i, j] = Inf
                continue
            end

            if rank(preproc_df(df[:, [i, j]])) != 4
                granger_mat[i, j] = Inf
                continue
            end
            
            #println(i, " ", j)
            _, signif = granger_cause(df[:, [i, j]])

            granger_mat[i, j] = signif ? 1 : 0
        end
    end

    return granger_mat
end

df = rand(300, 40)

begin
granger_conect(df)

replace_inf(x) = isinf(x) ? 0.0 : x

function granger_degree(x)
    return sum(replace_inf.(x)) / ((size(x)[2])^2 - size(x)[2] - sum(isinf.(x)))
end    

cor_w = 100

granger_ts = zeros(length(cor_w:size(data)[1]))

for t in cor_w:size(data)[1]
    println(t / size(data)[1])
    window = data[(t - cor_w + 1):t, :]
    mat = granger_conect(window)
    granger_ts[t - cor_w + 1] = granger_degree(mat)
end

na_share(data[(585 - cor_w + 1):585, [16,29]])

rank(remove_infs(data[(576 - cor_w + 1):576, [16,29]]))

df = data[(576 - cor_w + 1):576, :]

rank(preproc_df(data[(576 - cor_w + 1):576, [16, 29]]))

rank(rand(20, 2))

rank(remove_infs(df[:, [i, j]])) < 2

pair = [df[2:end, 1] add_lags(df[:, 1], 1)[:,2] add_lags(df[:, 2], 1)[:,2]]
pair = remove_infs(pair)
pair = [pair ones(size(pair, 1))]

df[(t - cor_w + 1):t, [16, 29]]
window[:, [16, 29]]

df = data[(t - cor_w + 1):t, :]
df = window
granger_cause(df[:, [15, 28]])
window
granger_conect(df[(t - cor_w + 1):t, :])
window[:, [16, 29]]



df_t = df[:, [15, 28]]

remove_infs(df[:, [15, 28]])

pair = [df_t[2:end, 1] add_lags(df_t[:, 1], 1)[:,2] add_lags(df_t[:, 2], 1)[:,2]]
pair = remove_infs(pair)
pair = [pair ones(size(pair, 1))]

β = (pair[:,2:end]'*pair[:,2:end])\pair[:,2:end]'*pair[:,1]
Σ = std_err(pair[:,2:end], pair[:,1], β)

return β[2], abs(β[2] / Σ[2]) > 1.96


    

File: data/granger_ts.py
import pandas as pd
import numpy as np
import statsmodels.api as sm
import yfinance as yf
from fredapi import Fred
from scipy.linalg import lstsq as sp_lstsq
import timeit
from scipy.linalg import solve

region = 'eu'

def na_share(df):
    return np.sum(df.isna().sum()) / (df.shape[0])

def std_err(X, y, beta):
    y = np.asarray(y)
    X = np.asarray(X)
    mse = np.mean((y - np.matmul(X, beta))**2)
    sigma_params = mse * np.linalg.inv(np.matmul(X.T, X))
    return np.sqrt(np.diag(sigma_params))

def granger_cause(df):
    pair = pd.concat([df.iloc[:, 0], df.iloc[:, 0:2].shift()], axis=1)\
        .dropna()

    # old and slow method:
    #model = sm.OLS(pair.iloc[:, 0], sm.add_constant(pair.iloc[:, 1:3]))
    #results = model.fit()
    
    pair['const'] = 1
    beta, _, _, _ = sp_lstsq(pair.iloc[:,[1,2,3]], pair.iloc[:,0], lapack_driver='gelsy', check_finite=False)
    sigma = std_err(pair.iloc[:,[1,2,3]], pair.iloc[:,0], beta)
    
    # return results.params[2], results.pvalues[2] < 0.05
    return beta[1], np.abs(beta[1] / sigma[1]) > 1.96 
    
def granger_mat(df):
    
    granger_mat = np.zeros((df.shape[1], df.shape[1]))
    
    for i in range(1, len(df.columns)):
        for j in range(1, len(df.columns)):
            
            if  i == j:
                granger_mat[i,j] = 0
                continue
            
            if (na_share(df.iloc[:,i]) > 0.95) or (na_share(df.iloc[:,j]) > 0.95):
                granger_mat[i,j] = np.nan
                continue
            
            _, signif = granger_cause(df.iloc[:, [i,j]])
            
            granger_mat[i,j] = 1 if signif else 0
            #print(df.columns[i], df.columns[j])
    return granger_mat  

fred = Fred(api_key='18c2830f79155831d5c485d84472811f')

if region == 'eu':
    print('region: EU')
    spread = fred.get_series('BAMLHE00EHYIOAS') # euro
    tickers = "EBO.DE RAW.DE KBC.BR CBK.DE DBK.DE NDA-SE.ST DANSKE.CO JYSK.CO SYDB.CO BBVA BKT.MC CABK.MC SAB.MC SAN.MC UNI.MC BNP.PA ACA.PA GLE.PA ALPHA.AT EUROB.AT ETE.AT TPEIR.AT OTP.BD A5G.IR BARC.L BIRG.IR BAMI.MI ISP.MI MB.MI BMPS.MI BPE.MI UCG.MI ABN.AS INGA.AS DNB.OL PKO.WA PEO.WA BCP.LS SEB-A.ST SHB-A.ST SWED-A.ST"
    banks_index = pd.read_excel("data/stoxx_banks.xlsx")\
                    .sort_index(ascending=False)\
                    .set_index('Date')
                    
    index = pd.DataFrame(yf.download("^STOXX", start="2000-01-01", group_by='tickers')['Open'])        
                    
elif region == 'us':
    print('region: US')
    spread = fred.get_series('BAMLC0A0CM') # usa
    tickers = "BAC BK BCS BMO COF SCHW C CFG DB GS JPM MTB MS NTRS PNC STT TD TFC UBS WFC ALLY AXP DFS FITB HSBC HBAN KEY MUFG PNC RF SAN"
    banks_index = pd.DataFrame(yf.download("^BKX", start="2000-01-01", group_by='tickers')['Open'])        
    index = pd.DataFrame(yf.download("^SPX", start="2000-01-01", group_by='tickers')['Open'])        

data_raw = yf.download(tickers, start="2000-01-01", group_by='tickers')

df_rets = data_raw.xs('Close', axis=1, level=1, drop_level=True)\
            .pct_change()\
            .iloc[3:, :]

# subtracting the index returns from the bank returns                
df_rets = df_rets\
            .loc[:banks_index.index[-1],:]\
            .sub(banks_index.pct_change().loc[df_rets.index[0]:,:], axis='columns', fill_value=0)\
            .iloc[:,0:-1]

granger = pd.DataFrame({'degree': []})

df_rets.to_csv("data/df_rets_granger.csv") 


#cor_w = 100
# for t in range(cor_w, len(df_rets)):
#     print(t/len(df_rets))
#     window = df_rets.iloc[(t - cor_w):t, :]
#     mat = granger_mat(window)
#     degree = np.nansum(mat) / (mat.shape[0] * mat.shape[1] - mat.shape[0])
#     granger = pd.concat([granger, pd.DataFrame({'degree': [degree]})])





      